### ISSUE TITLE
r.now() in secondary index functions

### ISSUE THREAD

**Issue by danielmewes** [rethinkdb/rethinkdb#5472]
@marshall007 noticed that you can currently create a secondary index function with `r.now` in it:
```js
table.indexCreate('now', (row) => r.now())
```

I assume that the `r.now()` is turned into a constant at the time where the index is created?

It seems to me like this should be forbidden, as it's more confusing than useful.

Pinging @Tryneus and @mlucy .

**Comment by mlucy** [comment]
I agree this is confusing and probably nobody is using it for its actual semantics.  We should probably forbid it.  (For what it's worth, though, these *are* the intended semantics: it's identical to `table.indexCreate('now'){|row| Time.now}`.)

**Comment by thelinuxlich** [comment]
#3583 and #5283 are slightly related

**Comment by danielmewes** [comment]
> For what it's worth, though, these are the intended semantics: it's identical to `table.indexCreate('now'){|row| Time.now}`.

The semantics are inconsistent though. This function is correctly rejected by RethinkDB as being non-deterministic:
```js
table.indexCreate('uuid', (row) => r.uuid())
```

We can't (easily) detect which client-side functions were used to generate the ReQL lambda. However by accepting the server-side function `r.now()` in a secondary index, we're leaking the implementation detail that `r.now` is transformed into something deterministic at compile time. I think that is bad.

**Comment by mlucy** [comment]
I don't see how the semantics are inconsistent.  They may be (and in fact probably are) sub-optimal, but they aren't inconsistent: `r.f(r.now, r.now) <=> x = Time.now; r.f(x, x)`.  `UUID`s are different because you get a different UUID every time you call it.

As an aside, maybe part of the problem is that `r.now()` is presented as a function in Python and JS?  In Ruby I right `r.now` with no parens and it makes it more obvious that it's a global value rather than a function that gets the current time.
